Add plumbing for passing credentials to devices

sdb9696 opened the pull request:
Add credentials classes and cli support to support future implementation of authenticating protocols.  See https://github.com/python-kasa/python-kasa/pull/477 for discussion

rytilahti commented on kasa/__init__.py:
```diff
@@ -42,4 +47,7 @@
     "SmartStrip",
     "SmartDimmer",
     "SmartLightStrip",
+    "SmartDeviceAuthenticationException",
```
```suggestion
    "AuthenticationException",
```

rytilahti commented on kasa/__init__.py:
```diff
@@ -13,9 +13,14 @@
 from kasa.emeterstatus import EmeterStatus
-from kasa.exceptions import SmartDeviceException
+from kasa.exceptions import (
+    SmartDeviceAuthenticationException,
```
```suggestion
    AuthenticationException,
```
Let's keep it simple, the unfortunate naming with "Smart" cam from the predecessor of this lib and was kept for compat reasons.

sdb9696 replied:
Done

rytilahti reviewed (commented):
Looks good to me! I think the smartdevice ctor change should also go into this PR, as that will allow writing some unittests to make sure it's all working as expected.

sdb9696 commented:
> Looks good to me! I think the smartdevice ctor change should also go into this PR, as that will allow writing some unittests to make sure it's all working as expected.

Now done. Also changed the default of username and password to None and removed the check in discovery for 255 as per comments on https://github.com/python-kasa/python-kasa/pull/477.  Also had to tweak the cli handling of the json parameter because tests failed if they appeared after the test_json_output because echo remained a noop

sdb9696 commented:
>>Looks good to me! I think the smartdevice ctor change should also go into this PR, as that will allow writing some unittests to make sure it's all working as expected.

>Now done. Also changed the default of username and password to None and removed the check in discovery for 255 as per comments on https://github.com/python-kasa/python-kasa/pull/477. Also had to tweak the cli handling of the json parameter because tests failed if they appeared after the test_json_output because echo remained a noop

Hi @rytilahti, any thoughts on the latest here?  It should have everything you wanted now.

rytilahti reviewed (commented):
Thanks for the ping, I've been busy and forgot to reply here earlier. Would you mind adding a parametrized test for the constructors of smartdevice classes to verify that the input parameters are stored as expected?

sdb9696 commented on kasa/cli.py:
```diff
@@ -158,13 +175,16 @@ async def cli(
         echo = _nop_echo
+    else:
+        # Set back to default is required if running tests with CliRunner
+        echo = _strip_rich_formatting(click.echo)
```
fixed

sdb9696 commented on kasa/cli.py:
```diff
@@ -330,6 +366,11 @@ async def state(dev: SmartDevice):
+        if dev.credentials.username is not None:
+            echo(f"\tUsername: {dev.credentials.username}")
+        if dev.credentials.password is not None:
+            echo(f"\tPassword: {dev.credentials.password}")
```
fixed

rytilahti replied:
There's no output anymore at all, but the cli test still tests for it.

sdb9696 replied:
The test is testing that cli passes the credentials to the device.  Because the credentials aren't in the standard state output anymore the test patches the state function so this works ok.

rytilahti replied:
Oh, I see, I somehow missed that. Let's get this merged so we can move forward!

sdb9696 commented on kasa/tests/test_cli.py:
```diff
@@ -121,3 +122,74 @@ async def test_json_output(dev: SmartDevice, mocker):
+        if "smartbulb" in type_.lower():
+            if "length" in sysinfo:  # strips have length
+                return "lightstrip"
+            return "bulb"
```
This is getting the --type parameter in order to test the constructor.  I've made it re-use existing code a bit better now

sdb9696 commented:
> Thanks for the ping, I've been busy and forgot to reply here earlier. Would you mind adding a parametrized test for the constructors of smartdevice classes to verify that the input parameters are stored as expected?

I think I have done this by passing in all the discovery fixtures and calling cli with the --type parameter.

rytilahti commented:
> I think I have done this by passing in all the discovery fixtures and calling cli with the --type parameter.

It's more important to test against the API as that's likely more consumed by external users than the cli tool. Tests for both are obviously better, but the API takes priority.

sdb9696 commented:
> > I think I have done this by passing in all the discovery fixtures and calling cli with the --type parameter.
> 
> It's more important to test against the API as that's likely more consumed by external users than the cli tool. Tests for both are obviously better, but the API takes priority.

Ok, so a test that creates an instance itself and then checks the credentials are there? With parameters of all the possible derived classes of SmartDevice?

rytilahti commented:
> Ok, so a test that creates an instance itself and then checks the credentials are there? With parameters of all the possible derived classes of SmartDevice?

Something like this should suffice, the idea is just to avoid breaking that API by chance:
```
@pytest.mark.parametrize("cls", [SmartDevice, SmartBulb, ...])
def test_device_class_ctors(cls): 
    host = "127.0.0.1"
    port = 1234
    credentials = ...
    dev = cls(host, port=port, credentials=credentials)
    assert dev.host = host
    assert dev.port = port
    assert dev.credentials = credentials
```
 I'm not sure which file it should go to, though...

sdb9696 commented:
> > Ok, so a test that creates an instance itself and then checks the credentials are there? With parameters of all the possible derived classes of SmartDevice?
> 
> Something like this should suffice, the idea is just to avoid breaking that API by chance:
> 
> ```
> @pytest.mark.parametrize("cls", [SmartDevice, SmartBulb, ...])
> def test_device_class_ctors(cls): 
>     host = "127.0.0.1"
>     port = 1234
>     credentials = ...
>     dev = cls(host, port=port, credentials=credentials)
>     assert dev.host = host
>     assert dev.port = port
>     assert dev.credentials = credentials
> ```
> 
> I'm not sure which file it should go to, though...

Ok so added this test to test_smartdevice.py.

rytilahti commented on kasa/cli.py:
```diff
@@ -158,13 +176,17 @@ async def cli(
             pass
 
         echo = _nop_echo
+    else:
+        # Set back to default is required if running tests with CliRunner
+        global _do_echo
+        echo = _do_echo
```
Could we avoid changes to echo handling by simply overriding it inside the `test_cli.py`? That would be the preferred way, imo.

rytilahti commented on kasa/cli.py:
```diff
@@ -51,6 +51,8 @@ def wrapper(message=None, *args, **kwargs):
 
 pass_dev = click.make_pass_decorator(SmartDevice)
 
+echo = _do_echo
```
Move the assignment directly after the definitions, and add a short one-line comment to describe why it's done like this. Then I think it's good to go, fixing it properly (if worth it) belongs to another PR and another time..

sdb9696 replied:
Done

rytilahti commented:
I don't currently have the energy nor the time to dig into that, so let's just add a comment, keep it as it is for now and let's get this merged :-)

sdb9696 commented:
> I don't currently have the energy nor the time to dig into that, so let's just add a comment, keep it as it is for now and let's get this merged (after fixing the test that does not really test the real output) :-)

That's done.  Re: the test that doesn't check the real output it does check that credentials when properly supplied are passed from the cli to the device so I think this should be good to go
